packages/test/test-end-to-end-tests/src/test/signalPerfTelemetry.spec.ts

```diff
@@ -0,0 +1,234 @@
+} from "@fluidframework/test-utils/internal";
+import sinon from "sinon";
+
+type ContainerRuntime = IContainerRuntimeBase &
+	IRuntime & {
+		clientId?: string | undefined;
+
+		emit(event: "signal"): void;
+	};
+
+const waitForSignals = async (
+	numSignals: number,
+	...signallers: { on(e: "signal", l: () => void): void }[]
+) =>
+	Promise.all(
+		signallers.map(async (signaller, index) =>
+			timeoutPromise(
+				(resolve) => {
+					let count = 0;
+					signaller.on("signal", () => {
+						count++;
+						if (count === numSignals) {
+							resolve();
+						}
+					});
+				},
+				{
+					durationMs: 2000,
+					errorMsg: `Signaller[${index}] Timeout`,
+				},
+			),
+		),
+	);
+
+describeCompat("Signal performance telemetry", "NoCompat", (getTestObjectProvider, apis) => {
+	let provider: ITestObjectProvider;
+	let dataObject: ITestFluidObject;
+	let container: IContainer;
+	let containerRuntime: ContainerRuntime;
+	let logger: MockLogger;
+
+	beforeEach("getTestObjectProvider", async () => {
+		provider = getTestObjectProvider();
+
+		logger = new MockLogger();
+
+		const testContainerConfig: ITestContainerConfig = {
+			fluidDataObjectType: DataObjectFactoryType.Test,
+			loaderProps: {
+				logger,
+			},
+		};
+		container = await provider.makeTestContainer(testContainerConfig);
+		dataObject = await getContainerEntryPointBackCompat<ITestFluidObject>(container);
+		containerRuntime = dataObject.context.containerRuntime as ContainerRuntime;
+
+		// need to be connected to send signals
+		if (container.connectionState !== ConnectionState.Connected) {
+			await new Promise((resolve) => container.once("connected", resolve));
+		}
+	});
+
+	describe("for all drivers", () => {
+		itExpects(
+			"SignalLatency telemetry after 100 broadcast signals",
+			[
+				{
+					eventName: "fluid:telemetry:ContainerRuntime:SignalLatency",
+					clientType: "interactive",
+				},
+			],
+			async () => {
+				// Send 100+ broadcast signals to trigger latency telemetry event
+				for (let i = 0; i < 101; i++) {
+					containerRuntime.submitSignal("signal", "test");
+				}
+				await waitForSignals(101, containerRuntime);
+			},
+		);
+
+		itExpects(
+			"SignalLost error event after missing signal is detected",
+			[
+				{
+					eventName: "fluid:telemetry:ContainerRuntime:SignalLost",
+					clientType: "interactive",
+				},
+			],
+			async () => {
+				const processSignalStub = sinon.stub(containerRuntime, "processSignal");
+
+				processSignalStub.callThrough();
+
+				containerRuntime.submitSignal("signal", "test");
+				await waitForSignals(1, containerRuntime);
+
+				processSignalStub.callsFake((message, local) => {
+					// Simulate a dropped signal
+					containerRuntime.emit("signal");
+				});
+
+				containerRuntime.submitSignal("signal", "test");
+				await waitForSignals(1, containerRuntime);
+
+				processSignalStub.callThrough();
+
+				containerRuntime.submitSignal("signal", "test");
+				await waitForSignals(1, containerRuntime);
```

Comment:
Needs comments. I don't know what the stubs are doing. I think the `callsFake` is meant to eat the signal, but why does it do anything? Why `containerRuntime.emit("signal")`?
Not a fan of injecting into the containerRuntime as requiring it to have a certain implementation. So, I would call this tech debt if used.
We have the service in our control, is there an option where we control how it responds? I feel like that level of manipulation is what we want - to be able to test DIS.
If not the service, then it seems like some testing with a mock connection layer in unit tests.

Reply:
`callsFake` function doesn't need to do anything, I just put the emit to work better with waitForSignals to know the signal reached the processSignal function (removed the emit now). The stub just spies on the containerRuntime object and just replacing the method behavior once it is hit https://sinonjs.org/releases/v18/stubs/ 

If by service you mean local server/tinlicious/routerlicious/afr code (e.g. dropping a signal once received on the server), then I don't believe we have that in our control in client-side testing, as client and server functionality are in separate workspaces and are tested separately.

I'll try looking into this 'mock connection layer' and stop testing against real services.